### DISCUSSION TITLE
Upsert without Primary Key - Changes Id

### DISCUSSION THREAD

**Discussion by tbarus** [supabase/supabase#25878] [Questions]
I have a table with a the following indexes/constraints:
CREATE UNIQUE INDEX lists_pkey ON public.list USING btree (id)

CREATE UNIQUE INDEX unique_items ON public.list USING btree (col1, col2, col3, col4, col5)

I have another source system that processes external data, and ONLY contains the Col1,2,3,4,5 data to insert, it does not have the ID from supa postgres.  So, i perform an upsert to ensure that 1 row is maintained with the latest data processed from my other source.  

THE ISSUE:
Upon upsert using the javascript SDK with onConflict cols 1-5, the ID column is updated to a new ID. So what was ID:1 with unique colums 1-5, now is ID: 325 with the same constraint columns 1-5, but different data,  ID:1 no longer exisits.  

WHAT I EXPECTED:
I expected upsert to perform update to every column i sent, none others, as long as it matched the onConflict col1-5.  So ID:1 with the new data in Cols 1-5.

Why is this happening and how to solve easily?

**Comment by GaryAustin1** [comment]
![image](https://github.com/supabase/supabase/assets/54564956/0e62778a-ec1d-4bcc-a276-66107904da8f)

**Reply by tbarus** [reply]
the upsert function works without the primary keys while using onConflict (ie in my case only using Cols1-5, without ID) against a unique constraint. However, the ID is changed. The ID changing is unexpected behavior.

**Reply by GaryAustin1** [reply]
Upsert does not work well with incrementing primary keys.
Upsert is "Insert on conflict do update".   
My GUESS as to what you are seeing is....
Insert occurs first and you did not provide an id, so Postgres uses the next available id for the insert and puts it in the row data.
That insert gets a conflict error based on your 5 columns.  Postgres then updates the row with all the data INCLUDING the new primary key id.

**Comment by khera** [comment]
I just wrote some code that does not send the primary key into the upsert and it seems to be leaving it alone. The ID is a UUID that is assigned a default by the DB. My upsert key is a pair of columns that have a unique constraint on them.

I think the docs need to be updated.
